lib/Parse/ParseExpr.cpp

```diff
@@ -3507,3 +3511,63 @@ Parser::parsePlatformVersionConstraintSpec() {
   return makeParserResult(new (Context) PlatformVersionConstraintAvailabilitySpec(
       Platform.getValue(), PlatformLoc, Version, VersionRange));
 }
+
+/// expr-pound-assert:
+///   '#assert' '(' expr ',' string_literal ')'
+ParserResult<Expr> Parser::parseExprPoundAssert() {
+  SyntaxParsingContext AssertContext(SyntaxContext,
+      SyntaxKind::PoundAssertExpr);
+
+  SourceLoc startLoc = consumeToken(tok::pound_assert);
+  SourceLoc endLoc;
+
+  auto errorAndSkipToEnd = [&]() -> ParserResult<Expr> {
+    skipUntilDeclStmtRBrace(tok::r_paren);
+    if (Tok.is(tok::r_paren)) {
+      endLoc = consumeToken();
+    } else {
+      endLoc = PreviousLoc;
+    }
+    return makeParserResult<Expr>(new (Context)
+                                      ErrorExpr(SourceRange(startLoc, endLoc)));
```

Comment:
This is nice, but I don't think every expression parsing should have recovery logic on its own. We usually let `parseExprList()` or `parseBraceItems()` to recover from parse error.
Any particular reason to recover here? If there's no reason, diagnose and return `makeParserError()` should be sufficient.

Reply:
No. Returning error makes `parseBraceItems()` [automatically recover](https://github.com/apple/swift/blob/b01776164569cece2706d048707be37e970ce585/lib/Parse/ParseStmt.cpp#L468-L477). You don't need to touch `parseBraceItems()`.